packages/dds/tree/src/core/tree/visitorUtils.ts

```diff
@@ -51,41 +49,46 @@ export function announceDelta(
 	detachedFieldIndex: DetachedFieldIndex,
 ): void {
 	const visitor = deltaProcessor.acquireVisitor();
-	visitDelta(delta, combineVisitors([visitor], [visitor]), detachedFieldIndex, latestRevision);
+	visitDelta(delta, combineVisitors([visitor]), detachedFieldIndex, latestRevision);
 	visitor.free();
 }
 
+export interface CombinedVisitor extends DeltaVisitor {
+	readonly type: "Combined";
+
+	readonly visitors: readonly CombinableVisitor[];
+}
+
+export type CombinableVisitor =
+	| (DeltaVisitor & { type?: never })
+	| AnnouncedVisitor
+	| CombinedVisitor;
+
 /**
+ * Combines multiple visitors into a single visitor.
  * @param visitors - The returned visitor invokes the corresponding events for all these visitors, in order.
- * @param announcedVisitors - Subset of `visitors` to also call {@link AnnouncedVisitor} methods on.
- * This must be a subset of `visitors`: if not the visitor will not have its path correctly set when the events are triggered.
- * When `visitors` are making changes to data, `announcedVisitors` can be used to get extra events before or after all the changes from all the visitors have been made.
- * This can, for example, enable visitors to have access to the tree in these extra events despite multiple separate visitors updating different tree related data-structures.
  * @returns a DeltaVisitor combining all `visitors`.
  */
-export function combineVisitors(
-	visitors: readonly DeltaVisitor[],
-	announcedVisitors: readonly AnnouncedVisitor[] = [],
-): DeltaVisitor {
-	{
-		const set = new Set(visitors);
-		for (const item of announcedVisitors) {
-			assert(set.has(item), 0x8c8 /* AnnouncedVisitor would not get traversed */);
-		}
-	}
+export function combineVisitors(visitors: readonly CombinableVisitor[]): CombinedVisitor {
+	const allVisitors = visitors.flatMap((v) => (v.type === "Combined" ? v.visitors : [v]));
+	const announcedVisitors = allVisitors.filter(
+		(v): v is AnnouncedVisitor => v.type === "Announced",
```

Comment:
In the latest version of TS, you don't even have to do the explicit narrow here! But we're not on that version yet :(